File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/inheritance/error_propagation/error_propagation.swift
import Inheritance
import KotlinStdlib
import Testing


enum MySwiftError: Error {
    case boom(String)
}

@Test
func swiftReverseThrowSurfacesToKotlin() throws {
    // A Swift override of a @Throws Kotlin method throws a Swift error; a Kotlin caller must observe
    // a thrown exception (not a trap), with the message preserved.
    class BadThrower: Thrower {
        override func mightThrow(prefix: String) throws -> String {
            throw MySwiftError.boom("swift-boom:" + prefix)
        }
    }
    #expect(callMightThrowCatching(t: BadThrower(), prefix: "p").contains("swift-boom:p"))

    // A Swift override that does NOT throw still returns normally through the reverse bridge.
    class GoodThrower: Thrower {
        override func mightThrow(prefix: String) throws -> String { prefix + "-swift-ok" }
    }
    #expect(callMightThrowCatching(t: GoodThrower(), prefix: "p") == "ok:p-swift-ok")

    // Original Kotlin implementation unaffected.
    #expect(callMightThrowCatching(t: Thrower(), prefix: "p") == "ok:p-kotlin-ok")
}

@Test
func kotlinExceptionRoundTripsThroughSwiftOverride() throws {
    // A Swift override calls `super` (a @Throws Kotlin method that
    // throws). Letting it propagate, the Kotlin caller must receive the ORIGINAL Kotlin exception —
    // identity and message preserved — not a generic re-wrap.
    class Propagator: SuperThrower {
        override func boom() throws -> String {
            return try super.boom() // Kotlin throws MyKotlinException("kotlin-boom")
        }
    }
    #expect(callBoomCatching(s: Propagator()) == "kotlin-exception:kotlin-boom")
    // Original Kotlin instance also yields the Kotlin exception.
    #expect(callBoomCatching(s: SuperThrower()) == "kotlin-exception:kotlin-boom")
}

@Test
func swiftErrorRoundTripsBackToSwift() throws {
    // A Swift override throws a Swift error; a @Throws Kotlin relay propagates it back
    // out to Swift, where it must arrive as the SAME Swift error type/value (forward SwiftError unwrap).
    class ThrowingRelayer: Relayer {
        override func relay() throws -> String { throw MySwiftError.boom("round-trip") }
    }
    do {
        _ = try callRelay(r: ThrowingRelayer())
        Issue.record("expected callRelay to throw")
    } catch let error as MySwiftError {
        guard case .boom(let message) = error else {
            Issue.record("unexpected MySwiftError case")
            return
        }
        #expect(message == "round-trip")
    }
}

@Test
func kotlinExceptionThrownBySwiftOverrideSurfacesToSwift() throws {
    // A Swift override throws a *Kotlin* exception object it constructed itself. It crosses into Kotlin via the
    // reverse bridge — where `kotlinThrowableRCRef` must hand over the original throwable instead of boxing it in
    // a `SwiftError` — the @Throws Kotlin relay propagates it, and the forward bridge must deliver it back to
    // Swift as the concrete exported `MyKotlinException`: not a `KotlinError` fallback, not a `SwiftError` re-wrap.
    class KotlinErrorRelayer: Relayer {
        override func relay() throws -> String {
            throw MyKotlinException(message: "swift-thrown-kotlin-error")
        }
    }
    do {
        _ = try callRelay(r: KotlinErrorRelayer())
        Issue.record("expected MyKotlinException to be thrown")
    } catch let error as MyKotlinException {
        #expect(error.message == "swift-thrown-kotlin-error")
    } catch {
        Issue.record("expected MyKotlinException, got \(type(of: error)): \(error)")
    }
}

@Test
func kotlinThrowsSwiftBackedExceptionSubclass() throws {
    // The thrown object is a *Swift* subclass of an exported Kotlin
    // `Exception`.
    class SwiftThrowableLeaf: ThrowableBranch {
        override func throwableValue() -> String { "swift>" + super.throwableValue() }
    }

    let value = SwiftThrowableLeaf(origin: "primary")

    do {
        try throwProvided(value: value)
        Issue.record("expected throwProvided to throw")
    } catch let error as SwiftThrowableLeaf {
        #expect(error === value)
    } catch {
        Issue.record("expected the original SwiftThrowableLeaf, got \(type(of: error))")
    }
}

File: native/swift/swift-export-standalone-integration-tests/simple/testData/execution/inheritance/error_propagation/error_propagation.kt
// KIND: STANDALONE
// FREE_COMPILER_ARGS: -opt-in=kotlin.native.internal.InternalForKotlinNative
// MODULE: Inheritance
// FILE: error_propagation.kt


// --- Throwing reverse bridges ---

class MyKotlinException(message: String) : RuntimeException(message)

// Open `@Throws` method with a non-Unit return and a parameter: a Swift override may throw a Swift
// error, which must surface to a Kotlin caller as a thrown exception (reverse error bridge).
open class Thrower {
    @Throws(Throwable::class)
    open fun mightThrow(prefix: String): String = prefix + "-kotlin-ok"
}

// Calls the (possibly Swift-overridden) method and reports what propagated back to Kotlin.
fun callMightThrowCatching(t: Thrower, prefix: String): String = try {
    "ok:" + t.mightThrow(prefix)
} catch (e: MyKotlinException) {
    "kotlin-exception:" + e.message
} catch (e: Throwable) {
    "throwable:" + (e.message ?: "?")
}

// A Kotlin super that throws: used to check that a Kotlin exception let-propagate by a Swift override
// (which called `super`) comes home to Kotlin as the ORIGINAL Kotlin exception (identity preserved).
open class SuperThrower {
    @Throws(Throwable::class)
    open fun boom(): String = throw MyKotlinException("kotlin-boom")
}

fun callBoomCatching(s: SuperThrower): String = try {
    s.boom()
} catch (e: MyKotlinException) {
    "kotlin-exception:" + e.message
} catch (e: Throwable) {
    "throwable:" + (e.message ?: "?")
}

// Round-trip: a Swift override throws a Swift error; this `@Throws` relay propagates it back out to
// Swift, where it must arrive as the SAME Swift error (forward SwiftError unwrap).
open class Relayer {
    @Throws(Throwable::class)
    open fun relay(): String = "kotlin-relay"
}

@Throws(Throwable::class)
fun callRelay(r: Relayer): String = r.relay()

// An exported Kotlin Throwable with an overridable member, so that a Swift subclass of it is a genuinely
// Swift-backed object (dynamic Kotlin class, patched slot) rather than a bare wrapper.
open class ThrowableBranch(val origin: String) : Exception("kotlin throwable: $origin") {
    open fun throwableValue(): String = "kotlin-throwable:$origin"
}

@Throws(Throwable::class)
fun throwProvided(value: ThrowableBranch): Unit = throw value
